lib/reporting/reports/enterprise_fee_summary/enterprise_fees_with_tax_report_by_producer.rb

```diff
@@ -0,0 +1,292 @@
+              {
+                tax_rate_id: tax_rate_id,
+                enterprise_fee_id: item[:enterprise_fee_id],
+                order: item[:order],
+              }
+            end
+          end
+        end
+
+        def join_supplier
+          proc do |item|
+            order = item[:order]
+            order.line_items.map(&:supplier_id).map do |supplier_id|
+              {
+                tax_rate_id: item[:tax_rate_id],
+                enterprise_fee_id: item[:enterprise_fee_id],
+                supplier_id: supplier_id,
+                order: order
+              }
+            end
+          end
+        end
+
+        def group_keys
+          proc do |hash|
+            [
+              hash[:tax_rate_id],
+              hash[:enterprise_fee_id],
+              hash[:supplier_id],
+              hash[:order].distributor_id,
+              hash[:order].order_cycle_id
+            ]
+          end
+        end
+
+        def change_root_to_order
+          proc do |_, v|
+            v.map!{ |item| item[:order] }
+          end
+        end
+
+        def columns
+          {
+            distributor: :distributor,
+            producer: :producer,
+            producer_tax_status: :producer_tax_status,
+            order_cycle: :order_cycle,
+            enterprise_fee_name: :enterprise_fee_name,
+            enterprise_fee_type: :enterprise_fee_type,
+            enterprise_fee_owner: :enterprise_fee_owner,
+            tax_category: :tax_category,
+            tax_rate_name: :tax_rate_name,
+            tax_rate: :tax_rate_amount,
+            total_excl_tax: :total_excl_tax,
+            tax: :tax,
+            total_incl_tax: :total_incl_tax
+          }
+        end
+
+        def rules
+          [
+            { group_by: :distributor },
+            { group_by: :producer },
+            {
+              group_by: :order_cycle,
+              summary_row: proc do |_key, items, _rows|
+                order_ids = items.flat_map(&:second).map(&:id).uniq
+
+                enterprise_fees = Spree::Adjustment.enterprise_fee
+                  .where(order_id: order_ids)
+                  .pluck(:id, :amount)
+                enterprise_fee_adjustment_ids = enterprise_fees.map(&:first)
+                included_tax = Spree::Adjustment.tax
+                  .where(order: order_ids)
+                  .where(included: true)
+                  .where(adjustable_type: 'Spree::Adjustment')
+                  .where(adjustable_id: enterprise_fee_adjustment_ids)
+                  .pluck("sum(amount)")
+                  .first || 0
+                tax = Spree::Adjustment.tax
+                  .where(order: order_ids)
+                  .where(adjustable_type: 'Spree::Adjustment')
+                  .where(adjustable_id: enterprise_fee_adjustment_ids)
+                  .pluck("sum(amount)")
+                  .first || 0
+                total_excl_tax = enterprise_fees.sum(&:second) - included_tax
```

Comment:
I stuck for a while to answer this feedback : )

Some enterprise fees are connected to `line items` and others are connected to `orders`.
To load all the enterprise fee tax related to an order, I need to use:
```ruby 
order.all_adjustments.tax.inclusive
            .where(adjustable_type: 'Spree::Adjustment')
            .where(adjustable_id: enterprise_fee_adjustment_ids_for_orders(order_ids))
```
If you check the methods used in this report to calculate the values for the columns, you'll see that I didn't load `order.all_adjustments`. 

Instead, I used `Spree::Adjustment.tax` to do the calculation on the database and fetch directly the result without loading all the data. For example:
```ruby
        def tax(query_result_row, all: false, included: nil)
          order_ids = orders(query_result_row).map(&:id)
          adjustment_ids = enterprise_fee_adjustemnt_ids(query_result_row)
          query = Spree::Adjustment.tax
          query = query.where(included: true) unless included.nil?
          query = query.where(originator_id: tax_rate_id(query_result_row)) unless all == true
          query = query.where(order_id: order_ids)
            .where(adjustable_type: 'Spree::Adjustment')
            .where(adjustable_id: adjustment_ids)
            .pluck("sum(amount)")

          query.first || 0
        end
```
This will run two  queries (if we don't count the query used to load the order in the main query):
1. one query to get adjustment_ids 
2. one query to get tax (after being already calculated)

In the other approach, I would need to run order.all_adjustments.tax.where(...) for every order which would require:
1. running more queries (1 query per order)
2. loading more data 
    a. will need more RAM
    b. query will be slower to fetch data
    c. calculation will be slower (will run on ruby) 
 

I tried the other approach, the code will look like 
```ruby
        def included_tax_for_order_ids(orders)
          enterprise_fee_adjustment_ids = enterprise_fee_adjustment_ids_for_orders(orders.map(&:id))
          orders.map do |order|
            order.all_adjustments.tax.inclusive.filter do |adjustment|
              adjustment.adjustable_type== 'Spree::Adjustment' &&
              enterprise_fee_adjustment_ids.includes?(adjustment.adjustable_id) 
            end.sum(&:amount)
          end.sum
```

What do you think?

Reply:
Wow, this is quite complex. I thought that the summary row would only show a simple aggregation of the data that is  already loaded in the line items, but it sounds like that is not the case.

I definitely don't want to suggest a solution that uses more memory and more queries, so it looks like your first solution is the best. Thanks for the thorough explanation!

Reply:
> I thought that the summary row would only show a simple aggregation of the data that is already loaded in the line items

This is correct but to avoid loading everything into the memory, I pushed the work to the database.

Reply:
Oh, I believe that all `LineItems` are still in memory when the summary is created (which can be quite a lot, but is a necessary result of this framework that we use). 
So I don't think directly using the `LineItem` objects to produce the aggregate is going to use any more memory, and would probably be as fast as using the DB to aggregate. But there might be more to this equation, and I don't think it's necessary to deliberate any further. This still works :D

Reply:
> I believe that all LineItems are still in memory when the summary is created

Yes, that's correct. the `LineItems` are loaded, but:

1. The data we're looking for is stored on the `Adjustment` model.  (but I didn't need to load them) 
2. All the `Adjustments` have a reference to the `order`, but not all the adjustment refers to the `LineItems`.

The second point means that we need to load the orders and from there we can load the adjustments.

Reply:
Ah, thanks for explaining further :)